[JSC] Add VectorMulHigh / VectorMulLow

Constellation opened the pull request:
#### 34349713b4cbbb41b5cb3114da286ac900b7277f
<pre>
[JSC] Add VectorMulHigh / VectorMulLow
<a href="https://bugs.webkit.org/show_bug.cgi?id=293998">https://bugs.webkit.org/show_bug.cgi?id=293998</a>
<a href="https://rdar.apple.com/152548112">rdar://152548112</a>

Reviewed by Justin Michaud.

This patch implements VectorMulHigh and VectorMulLow. wasm SIMD
extmul_high / extmul_low are well mapped to these B3 opcodes. Previously
we are using ExtractLow / ExtractHigh and VectorMul, but VectorMul is
really complicated and costly operation so it is significantly worse in
terms of performance. Having VectorMulHigh and VectorMulLow
significantly improves performance when it is used in particular on
ARM64.

* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::vectorMulHigh):
(JSC::MacroAssemblerARM64::vectorMulLow):
* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::vectorMulHigh):
(JSC::MacroAssemblerX86_64::vectorMulLow):
(JSC::MacroAssemblerX86_64::vectorUshl8):
(JSC::MacroAssemblerX86_64::vectorUshr8):
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::vpunpcklbw_rrr):
(JSC::X86Assembler::vpunpckhbw_rrr):
(JSC::X86Assembler::vpunpckhwd_rrr):
(JSC::X86Assembler::vpunpcklwd_rrr):
(JSC::X86Assembler::vpunpckldq_rrr):
(JSC::X86Assembler::vpunpckhdq_rrr):
(JSC::X86Assembler::vpmuldq_rrr):
(JSC::X86Assembler::vpmulhuw_rrr):
(JSC::X86Assembler::vpmulhw_rrr):
(JSC::X86Assembler::vpmuludq_rrr):
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/B3Opcode.cpp:
(WTF::printInternal): Deleted.
* Source/JavaScriptCore/b3/B3Opcode.h:
* Source/JavaScriptCore/b3/B3SIMDValue.h:
* Source/JavaScriptCore/b3/B3Validate.cpp:
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::effects const):
(JSC::B3::Value::key const):
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/b3/B3ValueInlines.h:
* Source/JavaScriptCore/b3/B3ValueKey.cpp:
(JSC::B3::ValueKey::materialize const):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/b3/testb3.h:
* Source/JavaScriptCore/b3/testb3_1.cpp:
(run):
* Source/JavaScriptCore/b3/testb3_7.cpp:
(testVectorMulHigh):
(testVectorMulLow):
* Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addSIMDExtmul):
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
(JSC::Wasm::OMGIRGenerator::addSIMDExtmul):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/WasmSIMD128.h: Added.

Canonical link: <a href="https://commits.webkit.org/295873@main">https://commits.webkit.org/295873@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/944c91c990c5143ae2c905bc9b62dba84f0e7e9b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106378 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/26127 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16525 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111579 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56974 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26794 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34630 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80803 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109382 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/21230 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96002 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61131 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/20721 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14101 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56414 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/99017 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90568 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14136 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114438 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/104995 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33516 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24708 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89872 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33880 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/92231 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89575 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34461 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12281 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/29110 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17238 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33441 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/38853 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/129307 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/33187 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35227 "Found 1 new JSC stress test failure: wasm.yaml/wasm/gc/ref-i31-eq.js.wasm-collect-continuously (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36540 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34785 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e404d15c9c4051f384bf6a56a7e6275d8b0d3a0d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46298

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/fa623f9a457d2a4e1a48df1ef318c90ecdcfb861. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46298

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/03a82c53b8967c7abc3ea1269ea6a24071739f29. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46298

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/cd606df7ab35f72593814cd82096950928355072. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46298

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e80581f94cbc585fbaf9968433cbb5c9c93d9398. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46298

Constellation committed:
[JSC] Add VectorMulHigh / VectorMulLow
https://bugs.webkit.org/show_bug.cgi?id=293998
rdar://152548112

Reviewed by Justin Michaud.

This patch implements VectorMulHigh and VectorMulLow. wasm SIMD
extmul_high / extmul_low are well mapped to these B3 opcodes. Previously
we are using ExtractLow / ExtractHigh and VectorMul, but VectorMul is
really complicated and costly operation so it is significantly worse in
terms of performance. Having VectorMulHigh and VectorMulLow
significantly improves performance when it is used in particular on
ARM64.

* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::vectorMulHigh):
(JSC::MacroAssemblerARM64::vectorMulLow):
* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::vectorMulHigh):
(JSC::MacroAssemblerX86_64::vectorMulLow):
(JSC::MacroAssemblerX86_64::vectorUshl8):
(JSC::MacroAssemblerX86_64::vectorUshr8):
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::vpunpcklbw_rrr):
(JSC::X86Assembler::vpunpckhbw_rrr):
(JSC::X86Assembler::vpunpckhwd_rrr):
(JSC::X86Assembler::vpunpcklwd_rrr):
(JSC::X86Assembler::vpunpckldq_rrr):
(JSC::X86Assembler::vpunpckhdq_rrr):
(JSC::X86Assembler::vpmuldq_rrr):
(JSC::X86Assembler::vpmulhuw_rrr):
(JSC::X86Assembler::vpmulhw_rrr):
(JSC::X86Assembler::vpmuludq_rrr):
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/B3Opcode.cpp:
(WTF::printInternal): Deleted.
* Source/JavaScriptCore/b3/B3Opcode.h:
* Source/JavaScriptCore/b3/B3SIMDValue.h:
* Source/JavaScriptCore/b3/B3Validate.cpp:
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::effects const):
(JSC::B3::Value::key const):
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/b3/B3ValueInlines.h:
* Source/JavaScriptCore/b3/B3ValueKey.cpp:
(JSC::B3::ValueKey::materialize const):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/b3/testb3.h:
* Source/JavaScriptCore/b3/testb3_1.cpp:
(run):
* Source/JavaScriptCore/b3/testb3_7.cpp:
(testVectorMulHigh):
(testVectorMulLow):
* Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addSIMDExtmul):
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
(JSC::Wasm::OMGIRGenerator::addSIMDExtmul):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/WasmSIMD128.h: Added.

Canonical link: https://commits.webkit.org/295873@main
